### PR TITLE
Add `isNormalForm` function

### PR DESCRIPTION
**Pull request by intricate**
`isNormalForm` enables us to determine whether an expression is fully evaluated.

Adapted from the `ghc-heap-view` package.

### CODE REVIEW

**Comment by erikd** [comment]
Addition of a simple test would be good. Basically something along the lines of:
```
prop_isNormalForm_correct :: Property
prop_isNormalForm_correct = do
  xs <- H.forAll $ Gen.list (Range.linear 0 10) Gen.int
  beforeForce <- isNormalForm xs
  afterForce <- isNormalFrom $ force xs
  H.assert  $ afterForce && not beforeFore
```
should do the trick. List of less than 2 elements may fail this test. If so, we should change the assertion rather than the generator.

File: cardano-prelude.cabal [hunk 1]
```diff
@@ -65,3 +65,52 @@ library
...
+                     , base16-bytestring
+                     , bytestring
+                     , canonical-json
+                     , cardano-prelude
+                     , containers
+                     , cryptonite
+                     , formatting
+                     , ghc-heap
+                     , hedgehog
+                     , hspec
```
Comment (erikd) [inline cardano-prelude.cabal]: Are using both `hspec` and `hedgehog` or was this just copy-pastaed ? Might be worth running `weeder` on this.

File: test/test.hs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  )
+where
+
+import Cardano.Prelude
+import Test.Cardano.Prelude
+
+import qualified Test.Cardano.Prelude.GHC.Heap
+
+main :: IO ()
+main = do
```
Comment (erikd) [inline test/test.hs]: Redundant `do`?

File: test/Test/Cardano/Prelude/GHC/Heap.hs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  , withTests
+  , annotateShow
+  )
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+-- | Property: Fully evaluated '[Int]'s are in normal form.
+prop_isNormalForm_correct :: Property
+prop_isNormalForm_correct = withTests 1000 $ property $ do
+  xs <- forAll $ Gen.list (Range.linear 0 1000) (Gen.int Range.constantBounded)
```
Comment (erikd) [inline test/Test/Cardano/Prelude/GHC/Heap.hs]: Lists of `1000` may be overkill. `100` may be better.

File: cardano-prelude.cabal (hunk 1 above)
Comment (intricate) [reply cardano-prelude.cabal]: `hspec` seems to be referenced from `Test.Cardano.Prelude.QuickCheck.Property`

Comment (intricate) [reply cardano-prelude.cabal]: `weeder` doesn't report any redundant build-depends in the test suite.

Comment (erikd) [reply cardano-prelude.cabal]: Fair enough!

File: src/Cardano/Prelude/GHC/Heap.hs [hunk 4]
```diff
@@ -0,0 +1,50 @@
+{-# LANGUAGE DoAndIfThenElse #-}
+
+-- | This code has been adapted from the module 'GHC.AssertNF' of the
+--   package 'ghc-heap-view' authored by Joachim Breitner.
```
Comment (mhuesch) [inline src/Cardano/Prelude/GHC/Heap.hs]: This is probably good enough, but I wonder if referencing a file / commit from the `ghc-heap-view` would make things even more clear. 

`ghc-heap-view` [has a BSD3 license](http://hackage.haskell.org/package/ghc-heap-view), and this [SO answer](https://softwareengineering.stackexchange.com/questions/234437/bsd-mit-license-changes-when-copy-pasting-a-function-from-another-bsd-mit-licens) seems to indicate citing the source is good enough. 

IANAL, though.

Comment (erikd) [reply src/Cardano/Prelude/GHC/Heap.hs]: I would add a link to the hackage page and the github repo. I don't think referencing a file or git commit is necessary.

**Pull request merged**